### PR TITLE
Implement save upload, extend __request method to allow for multiparts

### PR DESCRIPTION
**Pull request by janerjak**
Should probably be tested a bit further before merging.
Note that I exchanged
`auth=self.auth if self.auth else None`
to
`auth=self.auth`
as that should also work (?)

If there is anything to change, feel free to keep me updated

### CODE REVIEW

File: pyfactorybridge/__init__.py [hunk 1]
```diff
@@ -34,19 +36,47 @@ def __init__(self, address, password=None, token=None):
         else:
             logging.error("No password provided, some functions may not work.")
 
-    def __request(self, function, properties={}) -> dict[Any]:
+    def __build_request_data(
+        self, function: str, properties: dict | None = None
+    ) -> None:
```
Comment (Jayy001) [inline pyfactorybridge/__init__.py:41]: `-> dict`

File: pyfactorybridge/__init__.py [hunk 2]
```diff
@@ -167,8 +197,35 @@ def load_game(self, SaveName, EnableAdvancedGameSettings) -> dict[str, str]:
             },
         )
 
-    def upload_save_game(self, data) -> None:
-        raise NotImplementedError
+    def upload_save_game(
+        self,
+        save_file_path: str,
```
Comment (Jayy001) [inline pyfactorybridge/__init__.py:202]: Could this be renamed to `SaveFilePath` to keep it consistent?

**Review by Jayy001 — CHANGES_REQUESTED** [review]
Just some minor nitpicks - apart from that, looks great - thank you!

File: pyfactorybridge/__init__.py (hunk 2 above)
Comment (janerjak) [reply pyfactorybridge/__init__.py:202]: Sure, I thought you only named variables directly passed to the API in PascalCase and used snake_case for the rest. I mean all other parameters are also starting in lower case.

But if you insist, I can change that :)

File: pyfactorybridge/__init__.py (hunk 1 above)
Comment (janerjak) [reply pyfactorybridge/__init__.py:41]: Ah yes, I first only made this function use side-effects but changed it down the line and forgot that.

File: pyfactorybridge/__init__.py (hunk 2 above)
Comment (Jayy001) [reply pyfactorybridge/__init__.py:202]: Ah, no, good point - that's fine, we can keep it like that

**Comment by Jayy001** [comment]
Perfect, thanks very much - I'll merge this and get the downloading sorted too

**Comment by janerjak** [comment]
> Perfect, thanks very much - I'll merge this and get the downloading sorted too

Awesome, thanks for the effort!
I'm excited for using this library

**Review by Jayy001 — APPROVED** [review]

**Pull request merged**